distchan.go

```diff
@@ -22,45 +48,59 @@ import (
 // Note that the returned value's Start() method must be called before any
 // messages will be passed. This gives the user an opportunity to register
 // encoders and decoders before any data passes over the network.
-func NewServer(ln net.Listener, out, in interface{}) *Server {
-	return &Server{
-		ln:     ln,
-		outv:   reflect.ValueOf(out),
-		inv:    reflect.ValueOf(in),
-		closed: make(chan struct{}),
-		done:   make(chan struct{}),
-		logger: log.New(os.Stdout, "[distchan] ", log.Lshortfile),
+func NewServer(ln net.Listener, out, in interface{}) (*Server, error) {
+	if in != nil && reflect.ValueOf(in).Kind() != reflect.Chan {
+		return nil, ErrorInIsNotChannel
+	}
+	if out != nil && reflect.ValueOf(out).Kind() != reflect.Chan {
+		return nil, ErrorOutIsNotChannel
 	}
+	return &Server{
+		ln:      ln,
+		outv:    out,
+		inv:     in,
+		closed:  make(chan struct{}),
+		done:    make(chan struct{}),
+		logger:  log.New(os.Stdout, "[distchan] ", log.Lshortfile),
+		chconn:  make(chan clientConn),
+		chbroad: make(chan net.Conn), //non-buffered
+	}, nil
 }
 
 // Server represents a registration between a network listener and a pair
 // of channels, one for input and one for output.
 type Server struct {
 	ln                 net.Listener
-	inv, outv          reflect.Value
+	inv, outv          interface{}
```

Comment:
What's the reasoning for changing these from `reflect.Value` to `interface{}`?

Reply:
See http://www.gopl.io/.
You use reflect for long-time structs.
Reflect uses unsafe pointers.
If garbage collector realocate reflected data, this unsafe pointers would be point to unacessible memory.

Reply:
Okay, I think that's a probably a good enough argument to switch to shorter-lived `reflect.Value`s. However, the link you provided doesn't have any content about reflection available online, and I can't seem to find any other thoughts on long-lived reflection.

Reply:
this is a book talk about this problem, read it

Reply:
Yeah, but my point is that I don't have access to that book without purchasing it, which I'd rather not do just to answer this one question.

Reply:
what about torrent? ;)